pkg/ebpf/offsetguess.go

```diff
@@ -438,3 +346,129 @@ func guess(m *elf.Module, cfg *Config) error {
 	}
 	return nil
 }
+
+type eventGenerator struct {
+	listener net.Listener
+	lport    uint16
+	conn     net.Conn
+}
+
+func newEventGenerator() (*eventGenerator, error) {
+	// port 0 means we let the kernel choose a free port
+	addr := fmt.Sprintf("%s:0", listenIP)
+	l, err := net.Listen("tcp4", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	lport, err := strconv.Atoi(strings.Split(l.Addr().String(), ":")[1])
+	if err != nil {
+		l.Close()
+		return nil, err
+	}
+
+	go acceptHandler(l)
+	e := &eventGenerator{listener: l, lport: uint16(lport)}
+	return e, nil
+}
+
+// Generate an event for offset guessing
+func (e *eventGenerator) Generate(status *tracerStatus, expected *fieldValues) error {
+	// Are we guessing the IPv6 field?
+	if status.what == guessDaddrIPv6 {
+		expected.daddrIPv6 = generateRandomIPv6Address()
+
+		// For ipv6, we don't need the source port because we already guessed it doing ipv4 connections so
+		// we use a random destination address and try to connect to it.
+		expected.daddrIPv6 = generateRandomIPv6Address()
+		bindAddress := fmt.Sprintf("[%s]:9092", ipv6FromUint32Arr(expected.daddrIPv6))
+
+		// Since we connect to a random IP, this will most likely fail. In the unlikely case where it connects
+		// successfully, we close the connection to avoid a leak.
+		if conn, err := net.DialTimeout("tcp6", bindAddress, 10*time.Millisecond); err == nil {
+			conn.Close()
+		}
+
+		return nil
+	}
+
+	// Ensure v4 connection is up. The same connection is used for guessing all v4 offsets.
+	if e.conn == nil {
+		bindAddress := fmt.Sprintf("%s:%d", listenIP, expected.dport)
+		conn, err := net.Dial("tcp4", bindAddress)
+		if err != nil {
+			return fmt.Errorf("error dialing %q: %v", bindAddress, err)
+		}
+
+		e.conn = conn
+
+		// get the source port assigned by the kernel
+		sport, err := strconv.Atoi(strings.Split(conn.LocalAddr().String(), ":")[1])
+		if err != nil {
+			return fmt.Errorf("error converting source port: %v", err)
+		}
+
+		expected.sport = uint16(sport)
+	}
+
+	// This triggers the KProbe handler attached to `tcp_get_info`
+	_, err := tcpGetInfo(e.conn)
+	return err
+}
+
+func (e *eventGenerator) Close() {
+	if e.conn != nil {
+		e.conn.Close()
+	}
+
+	e.listener.Close()
+}
+
+func acceptHandler(l net.Listener) {
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			return
```

Comment:
Would it be useful to log here, perhaps ignoring the error when the listener is closed?

Reply:
Sure 👍

Reply:
Just to follow up on this, surprisingly there isn't a exported variable with the error you get when you close the listener (only `ErrNetClosing` on the internal `poll` package). 
I don't think it's worth the trouble of matching the error message, so I'll leave as it is.